### PR TITLE
Version output database tables

### PR DESCRIPTION
**Pull request by kyocum**
Created DBTarget class to allow users to make targets that represent output tables.   Users access the physical table name, and parameterize their queries to make or modify that table.   The DBTarget class creates unique table names under a supplied database schema.  Those names include the context and a prefix of the bundle uuid.

When users commit a bundle, we create a view that ties the physical table name to the tables virtual name (same schema and table but without the bundle uuid suffix).   Deletion of a bundle checks to see if the bundle is committed.  If not committed, we may safely delete any physical table.  If committed, the bundle may have been shared, i.e., there may be multiple outstanding references.   In this case, we check to see if the current view refers to this commit.   If not (a later commit now backs the view), then we are free to delete the physical table.   But if so, we require the user to 'force' the deletion.   

Note 1: We have also changed the database link in our protocol buffer definition so that it holds the DSN or data service name, and not logins or passwords.   We will need to support the ability for users to modify the dsn in the DBTarget to access tables if they use a different one.  

Note 2: Some of the specific queries are Vertica specific.   For example, to get the current database name, Vertica uses `select current_database()` as opposed to just `select database()`.   I have tried to use odbc information where possible.   However, future work will be to make this more generic. 

This PR also includes  a number of house-hold cleaning items.  Fixing some comments and removing some dead code.

### CODE REVIEW

File: disdat/pipe.py [hunk 1]
```diff
@@ -359,8 +353,28 @@ def run(self):
...
+
+            ASSUMES:  That we haven't actually updated the local DB with information on this bundle.
+
+            Returns:
+                None
+            """
             try:
                 shutil.rmtree(pce.path)
+
+                # if people create s3 files, s3 file targets, insdie of an s3 context,
```
Comment (barryhicks) [inline disdat/pipe.py]: Typo: “inside”

File: disdat/fs.py [hunk 2]
```diff
@@ -815,14 +823,17 @@ def commit(self, bundle_name, input_tags):
...
-        # Commit first in memory:
+        # Commit in memory:
         existing_tags = hfr.get_tags()
         existing_tags.update(tags)
         hfr.replace_tags(existing_tags)
 
         # Commit to disk:
         self.atomic_update_hframe(hfr)
 
+        # Address an commit issues with links in bundle:
```
Comment (barryhicks) [inline disdat/fs.py]: “Address and commit”

File: disdat/data_context.py [hunk 3]
```diff
@@ -946,24 +1056,17 @@ def copy_in_files(src_files, dst_dir):
...
+
+        This call works for src: dst pairs of the form:
+        local fs : managed local fs dir
+        local fs : managed s3 dir
+        s3       : managed s3 dir
+        s3       : managed local fs dir
+        db       : db
+
+        Note: We do not copy-in external tables to managed tables.
+        Note: While we convert
```
Comment (barryhicks) [inline disdat/data_context.py]: Seems like this note may have been truncated

File: disdat/db_target.py [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+    conn.close()
+
+
+def single_query(dsn, query, cursor=None):
+    """
+
+    Args:
+        dsn (unicode) : The database configuration dsn
+        query (unicode) : Raw query text
+        cursor:
```
Comment (barryhicks) [inline disdat/db_target.py]: and what is cursor?

File: disdat/db_target.py [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+    with get_cursor(dsn) as cursor:
+        for q in queries:
+            results.append(single_query(dsn, q, cursor))
+
+    return results
+
+
+def table_exists_vertica(dsn, schema, table):
+    """
+    Determine whether a table exists in a Vertica DataBase
```
Comment (barryhicks) [inline disdat/db_target.py]: does it only apply to Vertica?

File: disdat/db_target.py [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+def schema_exists_vertica(dsn, schema):
+    """
+    Determine whether a schema exists in a Vertica DataBase
+
+    Args:
+        dsn (unicode) : The db configuration to use
+        schema (unicode) : The schema holding table
+
+    Returns:
+        (bool) : True if table in db
```
Comment (barryhicks) [inline disdat/db_target.py]: True if schema in db

File: disdat/db_target.py [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+def drop_table_vertica(dsn, table, run=True):
+    """
+
+    Args:
+        dsn:
+        table:
+        run:
+
+
+    Returns:
```
Comment (barryhicks) [inline disdat/db_target.py]: would be useful to fill in the Args and Returns

File: disdat/db_target.py [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+def drop_view_vertica(dsn, table, run=True):
+    """
+
+    Args:
+        dsn:
+        table:
+        run:
+
+
+    Returns:
```
Comment (barryhicks) [inline disdat/db_target.py]: would be useful to fill in the Args and Returns

File: disdat/db_target.py [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+
+    return result
+
+
+def database_name_vertica(dsn):
+    """
+    SQL for getting the current database name
+
+    Returns:
+        (unicode):
```
Comment (barryhicks) [inline disdat/db_target.py]: would be useful to fill in the Args and Returns

**Comment by barryhicks** [comment]
Ken, I went through all the diffs. Since I'm not familiar with the code base, I'm not sure how useful my comments are, but there are a few small ones above.

**Comment by barryhicks** [comment]
I also noticed that when I ran the following command:
   dsdt context -d examples
I got the message:
   Disdat on context examples, please 'dsdt checkout <otherbranch>' before deleting.
I believe the checkout command is deprecated, so it would be good to adjust the above message. I think there's more than one such message in the code, so it would be useful to do a grep.

**Comment by barryhicks** [comment]
I did verify that when your branch is checked out, all the examples still run!

File: disdat/db_target.py (hunk 5 above)
Comment (kyocum) [reply disdat/db_target.py]: Yes.   Later, one possibility is to make the DBTarget be parameterized by SQL statements for other DB's.

File: disdat/db_target.py (hunk 6 above)
Comment (kyocum) [reply disdat/db_target.py]: fixed

File: disdat/db_target.py (hunk 7 above)
Comment (kyocum) [reply disdat/db_target.py]: Yes!  Done.

File: disdat/db_target.py (hunk 8 above)
Comment (kyocum) [reply disdat/db_target.py]: Done

File: disdat/db_target.py (hunk 9 above)
Comment (kyocum) [reply disdat/db_target.py]: Done

**Pull request merged**